coverage.py

```diff
@@ -262,11 +262,25 @@ def run_once_and_update_timestamp(self):
                         self.service_name, exc_info=e
                     )
                     progress.exception=traceback.format_exc()
+                    progress.finish=datetime.datetime.utcnow()
 
                 # The next run_once() call might raise an exception,
                 # so let's write the work to the database as it's
                 # done.
+                original_finish = progress.finish
                 self.finalize_timestampdata(progress)
+
+                # That wrote a value for progress.finish to the
+                # database, which is fine, but we don't necessarily
+                # want that value for progress.finish to stand. It
+                # might incorrectly make progress.is_complete appear
+                # to be True, making us exit the loop before we mean
+                # to.
+                if not progress.exception:
+                    progress.finish = original_finish
```

Comment:
I was really surprised by this bug. It means that no CoverageProvider will handle more than two batches (one batch for each value in `covered_status_lists`) on any given run. The first time it runs, `finalize_timestampdata` sets `progress.finish` for purposes of writing to the database. The loop thinks `finish` was set by `run_once`, and stops looking.

The only CoverageProvider where I would have noticed this problem is the search index one, and that one used a Monitor (removed in this branch), so it wouldn't have had this problem. It hasn't been a huge deal because our coverage providers run pretty frequently, but it does mean we've been getting coverage slower than we could have been.